core/src/main/scala/kafka/log/LogCleanerManager.scala

```diff
@@ -179,7 +179,11 @@ private[log] class LogCleanerManager(val logDirs: Seq[File],
         case (topicPartition, log) => // create a LogToClean instance for each
           try {
             val lastCleanOffset = lastClean.get(topicPartition)
+            val logStartOffset = log.logStartOffset
             val (firstDirtyOffset, firstUncleanableDirtyOffset) = cleanableOffsets(log, lastCleanOffset, now)
+            // update checkpoint for logs with invalid checkpointed offsets
```

Comment:
I agree with your comment on this PR that it is better to move the call to `updateCheckpoints` to `cleanableOffsets` / closer to the logic that resets the first dirty offset, since you know exactly when the offset was reset and there is no need to check again. Otherwise, we are opening more possibilities for future bugs if the logic changes but only one place gets updated.